lib/runtime-core/src/vm.rs

```diff
@@ -498,22 +498,49 @@ impl Ctx {
     }
 }
 
-enum InnerFunc {}
-/// Used to provide type safety (ish) for passing around function pointers.
-/// The typesystem ensures this cannot be dereferenced since an
-/// empty enum cannot actually exist.
+/// Represents a function pointer. It is mostly used in the
+/// `typed_func` module within the `wrap` functions, to wrap imported
+/// functions.
 #[repr(C)]
-pub struct Func(InnerFunc);
+pub struct Func {
+    _private: [u8; 0],
```

Comment:
This looks subtle. Are you relying on `#[repr(C)]` to produce a zero length array in the C API?

In C89 such things are UB (but a GCC extension), in C99 the same feature is available with different syntax, as flexible array members. Details: https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html

Reply:
The idea is to store a pointer that can't be deref (hence the `_private`). Do you have an other idea?

Reply:
I think the current solution for this kind of problem is to have an unsized `extern type`.

Reply:
@lachlansneff `std::mem::size_of::<[u8; 0]>()` and a struct containing it are 0.  That seems like that couldn't work unless there's some rule for keeping zero-sized types around at runtime that I don't know about.

It seems like `#[repr(transparent)]` with a private inner field should work here.

Let me know if you know of a reason why that can't work; you have a lot of low-level Rust expertise!

Reply:
Updated to `pub struct Func(pub(self) *mut c_void);`